substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs

```diff
@@ -596,18 +694,25 @@ where
 	/// After application, all already processed replacements are removed.
 	async fn apply_pending_tx_replacements(&self, view: Arc<View<ChainApi>>) {
 		let mut futures = vec![];
-		for replacement in self.pending_txs_replacements.read().values() {
-			let xt_hash = self.api.hash_and_length(&replacement.xt).0;
-			futures.push(self.replace_transaction_in_view(
-				view.clone(),
-				replacement.source.clone(),
-				replacement.xt.clone(),
-				xt_hash,
-				replacement.watched,
-			));
+		for replacement in self.pending_txs_tasks.read().values() {
+			match replacement.action {
```

Comment:
Are we protected against double execution of these actions?
I am imagining a scenario where we have [1] and this code here executing in parallel.
In [1] we already have processed then insertion of a `pending_txs_task`, but its not marked as processed yet. Then the codepath here could advance and process the freshly inserted task before it is marked processed, so it will not be removed at the end of `apply_pending_tx_replacements`.

Next time we arive here we will again process the same pending task and then remove it.

[1]: https://github.com/paritytech/polkadot-sdk/blob/2221d7a4ae0d897636ba465046d4bf94cd46e862/substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs#L834-L834

Reply:
The goal here is to avoid having unwanted transactions in newly cloned view which is being inserted. This whole cloning / building / updating / inserting is executed in context of maintain task. Here we execute tasks for the view that was newly created, and was not in the view_store when the tasks were scheduled.

The `processed` flag indicates that task was accomplished for every view in the view_store. Meaning that newly cloned view (which is not yet inserted) may contain undesired transactions, and tasks need to be executed.

What you say could happen. But it means that new (second) cloned view can be created while tasks are still not completed. Meaning that for this new view, the tasks also need to be executed.

Does it make sense?

Reply:
One more thought:

> Next time we arive here we will again process the same pending task and then remove it.

This is expected. Please note that processing will take place for another view (which was possibly cloned when not all tasks were yet processed for every view in view_store).